### PR TITLE
Use Mojolicious::Types to determine JSON content

### LINKED ISSUE #31: Definition of JSON is too narrow
I'm looking at a Dropbox v2 implementation.  I does not work with the plugin in its current state because of this:

```
sub _response_to_hash {
    my ( $self, $res ) = @_;
    return ( $res->headers->content_type =~ /^application\/json[;]?/ )
        ? $res->json
        : Mojo::Parameters->new( $res->body )->to_hash;
}
```

Dropbox returns JSON with a `content-type` of `text/javascript`.  I checked with #mojo on IRC and I was told that https://metacpan.org/pod/Mojolicious::Types would be a good way to solve this.

Perhaps by setting `$types->type(json => ['application/json', 'text/javascript']);` and maybe adding other `content-type` headers as they are required.

What are your thoughts on this?  Or would you rather just alter the regex?  I'm happy to send a pull request.

### PR DESCRIPTION
**Pull request by oalders**
Fixes #31

### CODE REVIEW

**Comment by oalders** [comment]
@jberger is this the right way to do this?

**Comment by hayajo** [comment]
:eyes:
Please wait until tomorrow as we will check.

**Comment by oalders** [comment]
Thanks @hayajo!

**Comment by oalders** [comment]
For what it's worth, this is what I was testing with: https://github.com/oalders/mojolicious-plugin-web-auth-site-dropboxv2/blob/master/lib/Mojolicious/Plugin/Web/Auth/Site/DropboxV2.pm

File: lib/Mojolicious/Plugin/Web/Auth/OAuth2.pm [hunk 1]
```diff
@@ -116,7 +117,10 @@ sub _ua {
 
 sub _response_to_hash {
     my ( $self, $res ) = @_;
-    return ( $res->headers->content_type =~ /^application\/json[;]?/ )
+    my $types = Mojolicious::Types->new;
+    $types->type(json => ['application/json', 'text/javascript']);
+    my $exts = $types->detect( $res->headers->content_type );
+    return ( $exts->[0] && $exts->[0] eq 'json' )
```
Comment (hayajo) [inline lib/Mojolicious/Plugin/Web/Auth/OAuth2.pm]: How about this?

```perl
return ( scalar(@$exts) && $exts->[0] eq 'json' )
```

Comment (oalders) [reply lib/Mojolicious/Plugin/Web/Auth/OAuth2.pm]: Good idea.  I've updated and force-pushed to the branch.

**Comment by hayajo** [comment]
LGTM! Thank you.
I will release a new version. Please wait a bit.

**Comment by oalders** [comment]
Thanks very much @hayajo!

File: lib/Mojolicious/Plugin/Web/Auth/OAuth2.pm [hunk 2]
```diff
@@ -116,7 +117,10 @@ sub _ua {
 
 sub _response_to_hash {
     my ( $self, $res ) = @_;
-    return ( $res->headers->content_type =~ /^application\/json[;]?/ )
+    my $types = Mojolicious::Types->new;
+    $types->type(json => ['application/json', 'text/javascript']);
+    my $exts = $types->detect( $res->headers->content_type );
+    return ( scalar(@$exts) && $exts->[0] eq 'json' )
```
Comment (jberger) [inline lib/Mojolicious/Plugin/Web/Auth/OAuth2.pm]: Does this line suffer from the parenthesis rule? I suspect it does.

Comment (jberger) [reply lib/Mojolicious/Plugin/Web/Auth/OAuth2.pm]: See https://perldoc.perl.org/perlop.html#Terms-and-List-Operators-(Leftward)

Comment (shadowcat-mst) [reply lib/Mojolicious/Plugin/Web/Auth/OAuth2.pm]: 16:57 <@mst> perlbot: deparse: return ( scalar(@$exts) && $exts->[0] eq 'json' )
16:57 < perlbot> mst: ( return ( scalar(@$exts) && ( $exts->[0] eq "json" ) ) 
                 );

**Pull request merged**